docs/idiomatic-maya-guide.md

```diff
@@ -0,0 +1,418 @@
+	}
+	nodeDiskMap, err := ac.getCandidateNode(listDisk)
+	if err != nil {
+		return nil, err
+	}
+	selectedDisk := ac.selectNode(nodeDiskMap)
+	return selectedDisk, nil
+}
+
+vs.
+
+// change the file path to below
+// pkg/nodeselect/v1alpha1/nodeselect.go
+
+// NodeDiskSelector selects and returns appropriate node
+//
+// Note: It is assumed that the information required to select is 
+// available in nodeselect instance
+//
+// Note: The return has to be an instance of node. This instance
+// should be defined in this package
+//
+// Note: Below might indicate a stripped off version, however it is
+// not. Code needs to be designed to place business logic in proper places.
+// A good code reflects in its Unit Testing as well as in its caller code.
+func (s *nodeselect) Get() (*node, error) {
+	d, err := s.getDisks()
+	if d == nil || len(d.Items) == 0 {
+		return nil, errors.Wrapf(err, "no disks provided")
+	}
+	dl := disk.NewList(disk.WithNames(d.Items)).Filter(disk.IsFree)
+	nl := node.NewList(node.WithDisks(dl.Item)).Filter(node.ContainsDisk)
+	return nl.Filter(node.IsPoolFeasible())
+}
+```
+
+### Good names are great but avoid repeating them
+
+```go
+type poolCreateConfig struct {
+  // the word algorithm is repeated
+	*algorithm.AlgorithmConfig
+}
+
+// vs.
+
+type poolCreateConfig struct {
+	*algorithm.Config
+}
+```
+
+```go
+poolconfig = &poolCreateConfig{
+  // here the word algorithm gets repeated
+	algorithm.NewAlgorithmConfig(spcGot),
+}
+
+// vs.
+
+poolconfig = &poolCreateConfig{
+	algorithm.NewConfig(spcGot),
+}
+```
+
+```go
+// the word new gets repeated; this is bad
+pool, err := newClientSet.NewCasPool(spcGot, poolconfig)
+
+// vs.
+
+p, err := cs.NewCasPool(spcGot, poolconfig)
+```
+
+### How do you name your packages?
+
+```go
+// Does below naming seem natural?
+// Does full path reflect what the logic intends to provide?
+// Does this adhere to golang's package naming convention?
+//  i.e. nodeSelect vs nodeselect
+//
+// refer - https://blog.golang.org/package-names
+// refer the other naming idioms mentioned in this doc
+pkg/algorithm/nodeSelect/v1alpha1
+
+// vs.
+
+pkg/nodeselect/v1alpha1
+```
+
+### Caller code should show the intent
+
+```go
+pool, err := newClientSet.NewCasPool(spcGot, poolconfig)
+
+// vs.
+
+// if caller needs a new instance of cas pool
+p, err := caspool.New(...)
+
+// vs. 
+// if caller needs to fetch from some service
+p, err := caspool.Get(..)
+```
+
+### Closed for modification yet open to extension
+- Avoid changing function signature like the one shown below
+- This can be caught by compilers, but is still a good one to sort during the design phase itself
+- Think how readability suffers, if you need to add a few more arguments to below function
+
+```diff
+- func (newClientSet *clientSet) NewCasPool(spc *apis.StoragePoolClaim) (*apis.CasPool, error) {
++ func (newClientSet *clientSet) NewCasPool(spc *apis.StoragePoolClaim, algorithmConfig *poolCreateConfig) (*apis.CasPool, error) {
+```
```

Comment:
This is not Open/Closed principle IMO. I am confused here.
Can arguments of existing function be added/removed depends on context.
Correct Me.